File: src/_cpp/TowerDefenseBase/scenes/gamegs/TowerPlace.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "ml/ImageManager.h"
#include "ml/Animation.h"
#include "ml/loadxml/xmlProperties.h"
#include "TowerPlace.h"
#include "support.h"
#include "ScoreCounter.h"
#include "configuration.h"
NS_CC_BEGIN





TowerPlace::TowerPlace( )
{}
TowerPlace::~TowerPlace( )
{}

bool TowerPlace::init( const TowerPlaseDef & def )
{
	do
	{
		CC_BREAK_IF( !Sprite::init() );
		_active = def.isActive;
		changeView();
		setPosition( def.position );
		return true;
	}
	while( false );
	return false;
}

void TowerPlace::changeView( )
{
	if( _active )
	{
		float duration = 0.5f + CCRANDOM_MINUS1_1()*0.1f;

		std::vector<std::string> frames;
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_01.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_02.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_03.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_04.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_05.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_06.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_07.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_08.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_09.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_10.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_11.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_12.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_13.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_14.png" );
		frames.push_back( k::resourceGameSceneFolder + "active_slot/active_slot_15.png" );
		auto anim = createAnimation( frames, duration );
		assert( anim );

		auto animate = RepeatForever::create( Animate::create( anim ) );
		auto action = animate ;

		runAction( action );
	}
	else
	{
		std::string texturename = k::resourceGameSceneFolder + "unactive_slot.png";
		auto frame = ImageManager::shared( ).spriteFrame( texturename );
		if( frame )
		{
			setSpriteFrame( frame );
			return;
		}
		xmlLoader::setProperty( this, xmlLoader::kImage, texturename );
	}
}

bool TowerPlace::checkClick( const Point & location, float & outDistance )
{
	outDistance = getPosition( ).getDistance( location );
	return checkRadiusByEllipse( location, getPosition(), 50 );
}

void TowerPlace::selected( )
{
	//auto s0 = ScaleTo::create( 0.5f, 1.5f );
	//auto s1 = ScaleTo::create( 0.5f, 1.0f );
	//auto action = RepeatForever::create( Sequence::create( s0, s1, nullptr ) );
	//runAction( action );
}

void TowerPlace::unselected( )
{
	//stopAllActions( );
	//setScale( 1 );
}

bool TowerPlace::getActive( void )
{
	return _active;
}

void TowerPlace::setActive( bool var )
{
	assert( _active == false );

	_active = true;
	changeView();
}



NS_CC_END

File: src/_eu/scenes/gamegs/TowerPlace.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.TowerPlaceDef = cc.Class.extend({
    position: null,
    isActive: null,
    ctor: function(){
        this.position = cc.p(0,0);
        this.isActive = true;
    }
});

EU.TowerPlace = cc.Sprite.extend(
{
    /** For Test Instance of */
    __TowerPlace : true,

    _cost : 0,
    _active : false,

    ctor: function( def )
    {
        this._super();
        this._active = def.isActive;
        this.changeView();
        this.setPosition( def.position );
        return true;
    },
    changeView: function( )
    {
        var image = "";
        if( this._active )
        {
            var duration = 0.5 + cc.randomMinus1To1()*0.1;

            //TODO: EU.Animation
            var frames = [];
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_01.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_02.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_03.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_04.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_05.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_06.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_07.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_08.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_09.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_10.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_11.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_12.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_13.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_14.png" );
            frames.push( EU.k.resourceGameSceneFolder + "active_slot/active_slot_15.png" );
            var anim = EU.Animation.createAnimation_t( frames, duration );
            EU.assert( anim );

            var animate = cc.repeatForever( cc.animate( anim ) );
            var action = animate ;

            this.runAction( action );
            image = EU.k.resourceGameSceneFolder + "active_slot/active_slot_01.png";
        }
        else
        {
            image = EU.k.resourceGameSceneFolder + "unactive_slot.png";
            var frame = EU.ImageManager.getSpriteFrame( image );
            if( frame )
            {
                this.setSpriteFrame( frame );
                return;
            }
            EU.xmlLoader.setProperty( this, EU.xmlKey.Image.name, image );
        }

    },
    
    checkClick: function(  location )
    {
        var result = {};
        result.clicked = EU.checkRadiusByEllipse( location, this.getPosition(), 50 );
        result.distance = cc.pDistance(this.getPosition(), location );
        return result;
    },
    selected: function( )
    {
        //var s0 = ScaleTo.create( 0.5f, 1.5f );
        //var s1 = ScaleTo.create( 0.5f, 1.0f );
        //var action = RepeatForever.create( Sequence.create( s0, s1, nullptr ) );
        //runAction( action );
    },
    unselected: function( )
    {
        //stopAllActions( );
        //setScale( 1 );
    },
    getActive: function( )
    {
        return this._active;
    },
    setActive: function()
    {
        EU.assert( this._active == false );
    
        this._active = true;
        this.changeView();
    }

});

    
    
   